src/main/java/com/maxmind/db/Decoder.java

```diff
@@ -93,7 +111,8 @@
             int targetOffset = (int) pointer;
             int position = buffer.position();
 
-            T o = cls.cast(cache.get(targetOffset, cls, cacheLoader));
+            CacheKey key = new CacheKey(targetOffset, cls, genericType);
```

Comment:
One possible optimization would be to  only create the cache key when we are using the cache. I'll defer to you and the profiling you did to determine whether it is worth it.

Reply:
Cool idea! I benchmarked doing this and it showed no real change though, so I'll leave it.